internal/rootcoord/garbage_collector.go

```diff
@@ -34,125 +27,84 @@ func newGarbageCollectorCtx(s *Core) *GarbageCollectorCtx {
+		baseStep:  baseStep{core: c.s},
+		pchans:    pChannels,
+		partition: partition,
+		ts:        ts,
+	})
+	redo.AddAsyncStep(&removeDmlChannelsStep{
+		baseStep:  baseStep{core: c.s},
+		pChannels: pChannels,
+	})
+	redo.AddAsyncStep(&removePartitionMetaStep{
+		baseStep:     baseStep{core: c.s},
+		collectionID: partition.CollectionID,
+		partitionID:  partition.PartitionID,
+		ts:           ts,
+	})
+
+	// err is ignored since no sync steps will be executed.
+	_ = redo.Execute(context.Background())
 }
 
-func (c *GarbageCollectorCtx) GcCollectionData(ctx context.Context, coll *model.Collection, ts typeutil.Timestamp) error {
-	msgPack := ms.MsgPack{}
-	baseMsg := ms.BaseMsg{
-		Ctx:            ctx,
-		BeginTimestamp: ts,
-		EndTimestamp:   ts,
-		HashValues:     []uint32{0},
-	}
-	msg := &ms.DropCollectionMsg{
-		BaseMsg: baseMsg,
-		DropCollectionRequest: internalpb.DropCollectionRequest{
-			Base: &commonpb.MsgBase{
-				MsgType:   commonpb.MsgType_DropCollection,
-				Timestamp: ts,
-				SourceID:  c.s.session.ServerID,
-			},
-			CollectionName: coll.Name,
-			CollectionID:   coll.CollectionID,
-		},
-	}
-	msgPack.Msgs = append(msgPack.Msgs, msg)
-	if err := c.s.chanTimeTick.broadcastDmlChannels(coll.PhysicalChannelNames, &msgPack); err != nil {
-		return err
-	}
-
-	// TODO: remove this after gc can be notified by rpc. Without this tt, DropCollectionMsg cannot be seen by
-	// 		datanodes.
-	return c.s.chanTimeTick.sendTimeTickToChannel(coll.PhysicalChannelNames, ts)
+func (c *GarbageCollectorCtx) GcCollectionData(ctx context.Context, coll *model.Collection, ts typeutil.Timestamp) (ddlTs Timestamp, err error) {
+	return c.s.ddlTsLockManager.NotifyCollectionGc(ctx, coll)
 }
 
-func (c *GarbageCollectorCtx) GcPartitionData(ctx context.Context, pChannels []string, partition *model.Partition, ts typeutil.Timestamp) error {
-	msgPack := ms.MsgPack{}
-	baseMsg := ms.BaseMsg{
-		Ctx:            ctx,
-		BeginTimestamp: ts,
-		EndTimestamp:   ts,
-		HashValues:     []uint32{0},
-	}
-	msg := &ms.DropPartitionMsg{
-		BaseMsg: baseMsg,
-		DropPartitionRequest: internalpb.DropPartitionRequest{
-			Base: &commonpb.MsgBase{
-				MsgType:   commonpb.MsgType_DropPartition,
-				Timestamp: ts,
-				SourceID:  c.s.session.ServerID,
-			},
-			PartitionName: partition.PartitionName,
-			CollectionID:  partition.CollectionID,
-			PartitionID:   partition.PartitionID,
-		},
-	}
-	msgPack.Msgs = append(msgPack.Msgs, msg)
-	if err := c.s.chanTimeTick.broadcastDmlChannels(pChannels, &msgPack); err != nil {
-		return err
-	}
-
-	// TODO: remove this after gc can be notified by rpc. Without this tt, DropCollectionMsg cannot be seen by
-	// 		datanodes.
-	return c.s.chanTimeTick.sendTimeTickToChannel(pChannels, ts)
+func (c *GarbageCollectorCtx) GcPartitionData(ctx context.Context, pChannels []string, partition *model.Partition, ts typeutil.Timestamp) (ddlTs Timestamp, err error) {
+	return c.s.ddlTsLockManager.NotifyPartitionGc(ctx, pChannels, partition)
```

Comment:
How about :
c.s.ddlTsLockManager.Lock()
allocate ts
do the NotfifyPartitionGC stuff as a method of GarbageCollector
c.s.ddlTsLockManager.updateLastTs(ts)
c.s.ddlTsLockManager.UnLock

Reply:
OK.